src/lib.rs

```diff
@@ -0,0 +1,238 @@
+// at your option. This file may not be copied, modified, or distributed except according to those
+// terms.
+
+#![feature(alloc_layout_extra)]
+
+use std::alloc::{alloc, Layout};
+use std::mem::{size_of, transmute};
+use std::time::Instant;
+
+const REPS: usize = 100;
+const VEC_SIZE: usize = 10000000;
+const VAL_NOREAD: usize = 0xdeadbeef;
+const VAL_WITHREAD: usize = 0xFEEDC0DE;
+
+#[inline(never)]
+fn time<F>(mut f: F)
+where
+    F: FnMut(),
+{
+    let before = Instant::now();
+    for _ in 0..REPS {
+        f();
+    }
+    let d = Instant::now() - before;
+    println!("{:?}", d.as_secs() as f64 + d.subsec_nanos() as f64 * 1e-9);
+}
+
+// The trait whose function we'll dynamically look up.
+trait T {
+    fn m(&self) -> usize;
+}
+
+// This is a simple helper trait that lets us cut a lot of our code in half below: it allows us to
+// instantiate SNoRead or SWithRead via a trait. We can't put this in T because T then wouldn't be
+// a valid trait object.
+trait N {
+    fn new() -> Self;
+}
+
+// SNoRead and SWithRead are the same size, but SNoRead.m() doesn't actually read the "_i"
+// attribute (hence the leading underscore) whereas SWithRead does read from its "i" attribute
+// (hence the lack of a leading underscore).
+struct SNoRead {
+    _i: usize,
+}
+
+impl T for SNoRead {
+    fn m(&self) -> usize {
+        VAL_NOREAD
+    }
+}
+
+impl N for SNoRead {
+    fn new() -> Self {
+        SNoRead { _i: VAL_NOREAD }
+    }
+}
+
+struct SWithRead {
+    i: usize,
+}
+
+impl T for SWithRead {
+    fn m(&self) -> usize {
+        self.i
+    }
+}
+
+impl N for SWithRead {
+    fn new() -> Self {
+        SWithRead { i: VAL_WITHREAD }
+    }
+}
+
+fn vec_normal<S: 'static + N + T>() -> Vec<Box<dyn T>> {
+    let mut v = Vec::<Box<dyn T>>::with_capacity(VEC_SIZE);
+    for _ in 0..VEC_SIZE {
+        v.push(Box::new(S::new()));
+    }
+    v
+}
+
+pub fn bench_normal_no_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_normal::<SNoRead>();
+    time(|| {
+        for e in &v {
+            assert_eq!(e.m(), VAL_NOREAD);
+        }
+    });
+}
+
+pub fn bench_normal_with_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_normal::<SWithRead>();
+    time(|| {
+        for e in &v {
+            assert_eq!(e.m(), VAL_WITHREAD);
+        }
+    });
+}
+
+fn vec_multiref<S: 'static + N + T>() -> Vec<*const dyn T> {
+    vec![Box::into_raw(Box::new(S::new())); VEC_SIZE]
+}
+
+pub fn bench_normal_multiref_no_read() {
+    assert_eq!(size_of::<Box<()>>(), size_of::<usize>());
+    assert_eq!(size_of::<Box<dyn T>>(), size_of::<usize>() * 2);
+    let v = vec_multiref::<SNoRead>();
+    time(|| {
+        for &e in &v {
+            assert_eq!(unsafe { (&*e).m() }, VAL_NOREAD);
```

Comment:
Talking of freeing, shouldn't you be making the elements back to boxes using `Box::from_raw`, thus ensuring they are freed?

Reply:
Only if we want to free the memory (as soon as you use `from_raw`, the recreated `Box` will free the memory when it's dropped).

Reply:
I know it's not essential, as the allocation doesn't happen inside the loop, but we probably should free our memory.

I'd expect a `Box::from_raw()` after the call to `time()`.

Reply:
I'm not too worried about that, because the benchmarks immediately terminate after the call.

Reply:
Granted. Again, I would have freed them anyway, but I won't insist.

Reply:
You win.